Python/bytecodes.c

```diff
@@ -2346,33 +2346,47 @@ dummy_func(
             assert(oparg & 1);
         }
 
-        // stack effect: (__0, __array[oparg] -- )
-        inst(CALL_BOUND_METHOD_EXACT_ARGS) {
-            DEOPT_IF(is_method(stack_pointer, oparg), CALL);
-            PyObject *function = PEEK(oparg + 1);
-            DEOPT_IF(Py_TYPE(function) != &PyMethod_Type, CALL);
-            STAT_INC(CALL, hit);
-            PyObject *self = ((PyMethodObject *)function)->im_self;
-            PEEK(oparg + 1) = Py_NewRef(self);
-            PyObject *meth = ((PyMethodObject *)function)->im_func;
-            PEEK(oparg + 2) = Py_NewRef(meth);
-            Py_DECREF(function);
-            GO_TO_INSTRUCTION(CALL_PY_EXACT_ARGS);
-        }
-
         inst(KW_NAMES, (--)) {
             assert(kwnames == NULL);
             assert(oparg < PyTuple_GET_SIZE(consts));
             kwnames = GETITEM(consts, oparg);
         }
 
-        // stack effect: (__0, __array[oparg] -- )
-        inst(CALL) {
+        // Cache layout: counter/1, func_version/2, min_args/1
+        // Neither CALL_INTRINSIC_1 nor CALL_FUNCTION_EX are members!
+        family(call, INLINE_CACHE_ENTRIES_CALL) = {
+            CALL,
+            CALL_BOUND_METHOD_EXACT_ARGS,
+            CALL_PY_EXACT_ARGS,
+            CALL_PY_WITH_DEFAULTS,
+            CALL_NO_KW_TYPE_1,
+            CALL_NO_KW_STR_1,
+            CALL_NO_KW_TUPLE_1,
+            CALL_BUILTIN_CLASS,
+            CALL_NO_KW_BUILTIN_O,
+            CALL_NO_KW_BUILTIN_FAST,
+            CALL_BUILTIN_FAST_WITH_KEYWORDS,
+            CALL_NO_KW_LEN,
+            CALL_NO_KW_ISINSTANCE,
+            CALL_NO_KW_LIST_APPEND,
+            CALL_NO_KW_METHOD_DESCRIPTOR_O,
+            CALL_METHOD_DESCRIPTOR_FAST_WITH_KEYWORDS,
+            CALL_NO_KW_METHOD_DESCRIPTOR_NOARGS,
+            CALL_NO_KW_METHOD_DESCRIPTOR_FAST,
+        };
+
+        // Stack is either
+        // [NULL, function, arg1, arg2, ...]
+        // or
+        // [method, self, arg1, arg2, ...]
+        // (Some args may be keywords, see KW_NAMES, which sets 'kwnames'.)
+        // It will be replaced with [result].
+        inst(CALL, (unused/1, unused/2, unused/1, thing1, thing2, unused[oparg] -- unused)) {
```

Comment:
> Maybe I missed the conversation during an earlier round of conversions, but why are the "unused" cache entries split up here (instead of just `unused/4`)?

So it's easier to see the correspondence between the cache layouts for various family members. The generator adds the sizes so there's no cost in the generated code. If it bothers you I can collapse them to `unused/4` where applicable.
 
> I think I also prefer @markshannon's suggestion of renaming `thing1` and `thing2` to `null_or_meth` and `func_or_self` throughout.

Yeah, I didn't go back to the opcodes I had already done using the `thing1/thing2` style (it was a bit of a death march to get the entire family done). The problem is that there's no really *consistent* way to name these throughout the whole family, since in some specializations it's always `method, self` or `NULL, callable`. I'll try to improve this though.

> Also, why are the arguments pulled from the stack manually rather than putting them in a named "array effect" (maybe I missed this during our discussion yesterday)? It would just be a pointer into the stack, so no copying overhead, right?

Hm, I need to revisit that too -- I did the right thing for most specializations, but initially I was hesitant to rewrite the code too much. I'll make things better.